File: 23-MergekSortedLists/Solution.java
/*
 * If we want to merge two sorted list, it would be easy. If we want 
 * to merge k sorted list, the only issue is that we need to compare
 * more times. So the solution is to use heap.
 *
 * Running time: O(nlogk)
 *
 * Chang Li at UC San Diego
 * Dec. 10, 2015
 */

public class Solution {
	class Node implements Comparable<Node> {
		public ListNode node;
		public Node(ListNode node) {
			this.node = node;
		}
		public int compareTo(Node n) {
			return this.node.val - n.node.val;
		}
	}

    public ListNode mergeKLists(ListNode[] lists) {
    	if(lists == null || lists.length == 0) {
            return null;
        }
        Queue<Node> heap = new PriorityQueue<Node>();
        for(ListNode n : lists) {
        	if(n != null) {
        		heap.offer(new Node(n));
	        }
        }
        ListNode dummy = new ListNode(0);
        ListNode pointer = dummy;
        while(heap.size() > 0) {
        	Node smallest = heap.poll();
        	if(smallest.node.next != null) {
        		heap.offer(new Node(smallest.node.next));
	        }
	        pointer.next = smallest.node;
	        pointer = pointer.next;
        }
        pointer.next = null;
        return dummy.next;
    }
}

File: 23-MergekSortedLists/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from queue import PriorityQueue
class Solution(object):
    def mergeKLists2(self, lists):
        dummy = ListNode(None)
        curr = dummy
        pq = []
        counter = 0
        for node in lists:
            if node:
                heapq.heappush(pq, (node.val, counter, node))
                counter += 1
        while len(pq) > 0:
            curr.next = heapq.heappop(pq)[2]
            curr = curr.next
            if curr.next:
                heapq.heappush(pq, (curr.next.val, counter, curr.next))
                counter += 1
        return dummy.next
      
    def mergeKLists(self, lists):
        dummy = ListNode(None)
        curr = dummy
        q = PriorityQueue()
        counter = 0
        for node in lists:
            if node:
                q.put((node.val, counter,node))
                counter += 1
        while q.qsize() > 0:
            curr.next = q.get()[2]
            curr = curr.next
            if curr.next:
                q.put((curr.next.val, counter, curr.next))
                counter += 1
        return dummy.next
      
"""
Complexity for this approach is O(klogk*n). We can also repeat -merge-2-lists for k, then
the complexity is O(k^2*n).
"""
